src/ripple/app/tests/ValidatorList_test.cpp

```diff
@@ -0,0 +1,228 @@
+
+            for (auto const& v : ephemeralValidators)
+                expect (static_cast<bool>(vl->member (v)) == isPresent (e, v));
+        }
+    }
+
+    void
+    testModification ()
+    {
+        testcase ("Insertion and Removal");
+
+        auto vl = make_ValidatorList ({}, {});
+
+        auto const v = randomNode ();
+
+        // Insert a permanent validator key
+        expect (vl->insertPermanentKey (v, "Permanent"));
+        {
+            auto member = vl->member (v);
+            expect (static_cast<bool>(member));
+            expect (member->compare("Permanent") == 0);
+        }
+        expect (!vl->insertPermanentKey (v, ""));
+        {
+            auto member = vl->member (v);
+            expect (static_cast<bool>(member));
+            expect (member->compare("Permanent") == 0);
+        }
+        expect (!vl->insertEphemeralKey (asPublicKey(v), "Ephemeral"));
+        {
+            auto member = vl->member (v);
+            expect (static_cast<bool>(member));
+            expect (member->compare("Permanent") == 0);
+        }
+        expect (!vl->removeEphemeralKey (asPublicKey(v)));
+        {
+            auto member = vl->member (v);
+            expect (static_cast<bool>(member));
+            expect (member->compare("Permanent") == 0);
+        }
+        // Deleting the key as permanent succeeds:
+        expect (vl->removePermanentKey (v));
+        {
+            auto member = vl->member (v);
+            expect (!static_cast<bool>(member));
+        }
+
+        // Insert an ephemeral validator key
+        expect (vl->insertEphemeralKey (asPublicKey(v), "Ephemeral"));
+        {
+            auto member = vl->member (v);
+            expect (static_cast<bool>(member));
+            expect (member->compare("Ephemeral") == 0);
+        }
+        // Inserting the same ephemeral key fails
+        expect (!vl->insertEphemeralKey (asPublicKey(v), ""));
+        {
+            auto member = vl->member (v);
+            expect (static_cast<bool>(member));
+            expect (member->compare("Ephemeral") == 0);
+        }
+        // Inserting the same key as ephemeral fails:
+        expect (!vl->insertPermanentKey (v, "Permanent"));
+        {
+            auto member = vl->member (v);
+            expect (static_cast<bool>(member));
+            expect (member->compare("Ephemeral") == 0);
+        }
+        // Deleting the key as permanent fails:
+        expect (!vl->removePermanentKey (v));
+        {
+            auto member = vl->member (v);
+            expect (static_cast<bool>(member));
+            expect (member->compare("Ephemeral") == 0);
+        }
+        // Deleting the key as ephemeral fails:
+        expect (vl->removeEphemeralKey (asPublicKey(v)));
+        {
+            auto member = vl->member (v);
+            expect (!static_cast<bool>(member));
+        }
+    }
+
+public:
+    void
+    run() override
+    {
+        testMembership ();
+        testModification ();
```

Comment:
I would like to suggest an additional test that mocks up a `Config` object with `validations` populated so that `loadValidators` can get exercised / covered. It's a glaring block of red on the coverage report.